scripts/calculate_udf_formulas.py

```diff
@@ -0,0 +1,474 @@
+#!/usr/bin/env python
+
+import logging
+import re
+from argparse import ArgumentParser
+from datetime import datetime as dt
+
+from genologics.config import BASEURI, PASSWORD, USERNAME
+from genologics.entities import Artifact, Process
+from genologics.lims import Lims
+
+from scilifelab_epps.utils.formula import (
+    fmol_to_ng,
+    ng_to_fmol,
+    ng_ul_to_nM,
+    nM_to_ng_ul,
+)
+from scilifelab_epps.utils.udf_tools import fetch_last, get_art_tuples
+from scilifelab_epps.wrapper import epp_decorator
+
+DESC = """Script to perform UDF calculations on input-output level by reading equations
+with special syntax from a step UDF text field.
+
+The whole idea is to have a single calculation script whose behavior can be customized
+on a step-by-step basis in the front-end configuration.
+
+The text field can contain multiple equations on one line each. Lines prefixed with '#'
+will be ignored as comments.
+
+Syntax explained:
+
+    - UDF placeholders
+
+        A special string referencing a UDF name in an artifact or step, and whether to fetch
+        it recursively. Can also be supplied as a prioritized list of UDF names.
+
+        E.g.
+            inp['foo']      step input artifact UDF 'foo'
+            _outp['bar']    artifact UDF 'bar', fetched recursively from step output
+                            artifact
+            step['A', 'B']  step UDF 'A', or 'B' if 'A' is not found
```

Comment:
Can this also be recursive? But that probably doesn't make any sense?

Reply:
It can, but it will be a "depth-first" search, i.e. will only try the 2nd UDF if the 1st could not be found recursively.

Reply:
So using the underscore is ok here as well as in `_step['A', 'B']`?

Reply:
Sorry, recursive search for prioritized lists of UDFs works, but recursive search doesn't work at all for step placeholders, only for artifacts. There's a specific error message for it